### PR TITLE
Cleaning TrainCatalog and RecommenderCatalog

### LINKED ISSUE #2972: TrainCatalog and RecommenderCatalog need cleaning
There are some methods in the catalogs that do not respect the naming conventions that we are enforcing in the trainers and the transforms. 

The methods that should be fixed are the `Evaluate` and `CrossValidate` in particular.

We should rename the parameters accordingly.

### PR DESCRIPTION
**Pull request by artidoro**
Fixes #2972.

In this PR I clean `TrainCatalog` (commit 1) and `RecommenderCatalog` (commit 2).
The other commits are simply the adjustments in the code that need to be done to compile.

### CODE REVIEW

File: test/Microsoft.ML.Tests/Scenarios/Api/CookbookSamples/CookbookSamplesDynamicApi.cs [hunk 1]
```diff
@@ -119,7 +119,7 @@ private void TrainRegression(string trainDataPath, string testDataPath, string m
                 hasHeader: true);
 
             // Calculate metrics of the model on the test data.
-            var metrics = mlContext.Regression.Evaluate(model.Transform(testData), label: "Target");
+            var metrics = mlContext.Regression.Evaluate(model.Transform(testData), labelColumnName: "Target");
```
Comment (abgoswam) [inline test/Microsoft.ML.Tests/Scenarios/Api/CookbookSamples/CookbookSamplesDynamicApi.cs]: might have to update cookbook.md file

File: src/Microsoft.ML.Data/TrainCatalog.cs [hunk 2]
```diff
@@ -558,41 +558,41 @@ internal RegressionTrainers(RegressionCatalog catalog)
...
+        /// <param name="samplingKeyColumname">Name of a column to use for grouping rows. If two examples share the same value of the <paramref name="samplingKeyColumname"/>,
         /// they are guaranteed to appear in the same subset (train or test). This can be used to ensure no label leakage from the train to the test set.
         /// If <see langword="null"/> no row grouping will be performed.</param>
         /// <param name="seed">Seed for the random number generator used to select rows for cross-validation folds.</param>
         /// <returns>Per-fold results: metrics, models, scored datasets.</returns>
         public CrossValidationResult<RegressionMetrics>[] CrossValidate(
-            IDataView data, IEstimator<ITransformer> estimator, int numFolds = 5, string labelColumn = DefaultColumnNames.Label,
-            string samplingKeyColumn = null, int? seed = null)
+            IDataView data, IEstimator<ITransformer> estimator, int numberOfFolds = 5, string labelColumnName = DefaultColumnNames.Label,
+            string samplingKeyColumname = null, int? seed = null)
```
Comment (abgoswam) [inline src/Microsoft.ML.Data/TrainCatalog.cs]: >samplingKeyColumname [](start = 19, length = 20)

ColumnName

Comment (artidoro) [reply src/Microsoft.ML.Data/TrainCatalog.cs]: Thank you for catching this!

File: test/Microsoft.ML.Tests/Scenarios/Api/CookbookSamples/CookbookSamplesDynamicApi.cs (hunk 1 above)
Comment (artidoro) [reply test/Microsoft.ML.Tests/Scenarios/Api/CookbookSamples/CookbookSamplesDynamicApi.cs]: Thanks for pointing out!!

File: src/Microsoft.ML.Recommender/RecommenderCatalog.cs [hunk 3]
```diff
@@ -98,42 +98,42 @@ internal RecommendationTrainers(RecommendationCatalog catalog)
...
+        /// Then evaluate each sub-model against <paramref name="labelColumnName"/> and return metrics.
         /// </summary>
         /// <param name="data">The data to run cross-validation on.</param>
         /// <param name="estimator">The estimator to fit.</param>
-        /// <param name="numFolds">Number of cross-validation folds.</param>
-        /// <param name="labelColumn">The label column (for evaluation).</param>
-        /// <param name="stratificationColumn">Optional name of the column to use as a stratification column. If two examples share the same value of the <paramref name="stratificationColumn"/>
+        /// <param name="numberOfFolds">Number of cross-validation folds.</param>
+        /// <param name="labelColumnName">The label column (for evaluation).</param>
+        /// <param name="samplingKeyColumn">Optional name of the column to use as a stratification column. If two examples share the same value of the <paramref name="samplingKeyColumn"/>
```
Comment (wschin) [inline src/Microsoft.ML.Recommender/RecommenderCatalog.cs]: >samplingKeyColumn [](start = 25, length = 17)

Maybe FoldIndexColumnName or FoldColumnName?

File: src/Microsoft.ML.Recommender/RecommenderCatalog.cs [hunk 4]
```diff
@@ -98,42 +98,42 @@ internal RecommendationTrainers(RecommendationCatalog catalog)
...
-        /// If the <paramref name="stratificationColumn"/> is not provided, the random numbers generated to create it, will use this seed as value.
+        /// <param name="seed">Optional parameter used in combination with the <paramref name="samplingKeyColumn"/>.
+        /// If the <paramref name="samplingKeyColumn"/> is not provided, the random numbers generated to create it, will use this seed as value.
         /// And if it is not provided, the default value will be used.</param>
         /// <returns>Per-fold results: metrics, models, scored datasets.</returns>
         public CrossValidationResult<RegressionMetrics>[] CrossValidate(
-            IDataView data, IEstimator<ITransformer> estimator, int numFolds = 5, string labelColumn = DefaultColumnNames.Label,
-            string stratificationColumn = null, int? seed = null)
+            IDataView data, IEstimator<ITransformer> estimator, int numberOfFolds = 5, string labelColumnName = DefaultColumnNames.Label,
+            string samplingKeyColumn = null, int? seed = null)
```
Comment (wschin) [inline src/Microsoft.ML.Recommender/RecommenderCatalog.cs]: >samplingKeyColumn [](start = 19, length = 17)

Would you consider FoldColumnName?

Comment (artidoro) [reply src/Microsoft.ML.Recommender/RecommenderCatalog.cs]: Wei-Sheng suggested partitionColumnName`, which I think reflects the role of the column better. 

---
In reply to: [266177816](https://github.com/dotnet/machinelearning/pull/2973#discussion_r266177816) [](ancestors = 266177816)

Comment (Ivanidzo4ka) [reply src/Microsoft.ML.Recommender/RecommenderCatalog.cs]: https://github.com/dotnet/machinelearning/pull/2537
https://github.com/dotnet/machinelearning/issues/2536
We had discussion regarding name of this column.
If you want to change it again, I would advice ask opinion of @rogancarr  and @TomFinley

Comment (TomFinley) [reply src/Microsoft.ML.Recommender/RecommenderCatalog.cs]: So, I also considered in the issues @Ivanidzo4ka linked the merits of "partition" in my own thinking, though I did not write it down (as I had internally rejected it). I will now write down why I rejected it *here*, and you can see if you agree or disagree.

The major objection is that, if I look forward, I anticipate a future where this same sort of key is useful in *any* sampling utilities. (There are multiple ways to sample a dataset than simple partitionings!) For the sake of that more forward looking perspective, I wanted it named something else. And we ultimately settled on "sampling key."

So I would prefer if we did not act on this suggestion.

Comment (TomFinley) [reply src/Microsoft.ML.Recommender/RecommenderCatalog.cs]: Incidentally I wouldn't consider it a disaster if we did this. (Keeping it as stratification would have been a disaster! 😄) It merely is slightly misleading I think. But of course I also see @wschin's point.

Comment (artidoro) [reply src/Microsoft.ML.Recommender/RecommenderCatalog.cs]: Thank you @Ivanidzo4ka, and @TomFinley for  your comments on this, I am reverting the commit to leave the name of `samplingKeyColumnName` as is.

**Pull request merged**